tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,284 @@
+
+  # Test for verifying peers joining a topic using `JOIN(topic)`
+  asyncTest "handle JOIN topic and mesh is updated":
+    let topic = "test-join-topic"
+
+    # Initialize the GossipSub system and simulate peer connections
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    # Simulate peer joining the topic using commonSubscribe
+    commonSubscribe(@[gossipSub], topic, voidTopicHandler)
+
+    # Check that peers are added to the mesh and the topic is tracked
+    check gossipSub.mesh[topic].len == 5
+    check gossipSub.topics.contains(topic)
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test for verifying peers leaving a topic using `LEAVE(topic)`
+  asyncTest "multiple peers join and leave topic simultaneously":
+    let
+      numberOfNodes = 6
+      topic = "foobar"
+    var nodes = newSeq[TestGossipSub]()
+
+    # Initialize each node and start the switch
+    for i in 0 ..< numberOfNodes:
+      let (gossipSub, _) = setupGossipSub(topic, 5)
+      nodes.add(gossipSub) # Add the gossipSub instance to the sequence
+      await gossipSub.switch.start()
+
+    # Subscribe all nodes to the topic using the commonSubscribe method
+    commonSubscribe(nodes, topic, voidTopicHandler)
+
+    # Allow time for subscription propagation
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # Ensure that all nodes are subscribed to the topic
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = nodes[i]
+      doAssert currentGossip.gossipsub.hasKey(topic), "Node is not subscribed to the topic"
+
+    # Allow time for mesh stabilization
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # Print the mesh size for all nodes before unsubscription
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = nodes[i]
+      echo "Node ", i, " mesh size: ", currentGossip.mesh.getOrDefault(topic).len
+
+    # Expected number of peers in the mesh
+    let expectedNumberOfPeers = numberOfNodes - 1
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = nodes[i]
+      check:
+        currentGossip.gossipsub[topic].len == expectedNumberOfPeers
+        currentGossip.mesh[topic].len == expectedNumberOfPeers
+        currentGossip.fanout.len == 0
+
+    # Simulate unsubscription of 3 peers
+    let peersToUnsubscribe = nodes[0].mesh[topic].toSeq()[0 .. 2]
+
+    # Unsubscribe these peers from all the nodes
+    for node in nodes:
+      for peer in peersToUnsubscribe:
+        node.unsubscribe(topic, voidTopicHandler)
+        echo "Unsubscribing peer: ", peer.peerId
+
+    # Allow time for heartbeat to adjust the mesh
+    await sleepAsync(3 * DURATION_TIMEOUT)
+
+    # Check the mesh size for the first node after unsubscription
+    let firstNodeGossip = nodes[0]
+    echo "Mesh size after unsubscription: ",
+      firstNodeGossip.mesh.getOrDefault(topic).len
+    doAssert firstNodeGossip.mesh.getOrDefault(topic).len == 3,
+      "Expected 3 peers to remain in the mesh"
```

Comment:
Just as you did before, I'd check the `gossipsub` and `mesh` (and `fanout`, if you want) are in the expected state, not just the `mesh`. As those two (`gossipsub` and `mesh`) are directly involved in the subscription/unsubscription you're testing).